unit/src/HitTests.C

```diff
@@ -0,0 +1,153 @@
+        << test.input << "'";
+  }
+}
+
+TEST(HitTests, PassCases)
+{
+  PassFailCase cases[] = {
+      {"valid special field chars", "hello_./:<>-+world=foo"},
+      {"comment in field", "hello=wo#rld"},
+      {"bad number becomes string", "foo=4.2abc"},
+      {"empty section close", "[hello] []"},
+      {"dotdotslash section close", "[hello] [../]"},
+      {"no whitespace between headers/footers", "[hello][../]"},
+      {"no whitespace between headers/footers", "[hello][]"},
+      {"no whitespace between headers/footers", "[hello][../]"},
+      {"no whitespace with sections and fields", "[hello][world]foo=bar[]baz=42[]"},
+      {"no leading ./ in sub-block", "[hello] [world] [] []"},
+  };
+
+  for (size_t i = 0; i < sizeof(cases) / sizeof(PassFailCase); i++)
+  {
+    auto test = cases[i];
+    try
+    {
+      hit::parse("TESTCASE", test.input);
+    }
+    catch (hit::Error & err)
+    {
+      FAIL() << "case " << i + 1 << " FAIL (" << test.name
+             << "): unexpected parser error on valid input '" << test.input << "': " << err.what();
+    }
+  }
+}
+
+std::string
+strkind(hit::Field::Kind k)
+{
+  if (k == hit::Field::Kind::String)
+    return "String";
+  else if (k == hit::Field::Kind::Bool)
+    return "Bool";
+  else if (k == hit::Field::Kind::Int)
+    return "Int";
+  else if (k == hit::Field::Kind::Float)
+    return "Float";
+  else if (k == hit::Field::Kind::None)
+    return "None";
+  else
+    return "Unknown";
+}
+
+struct ExpandCase
+{
+  std::string name;
+  std::string input;
+  std::string key;
+  std::string val;
+  hit::Field::Kind kind;
+};
+
+TEST(ExpandWalkerTests, All)
+{
+  ExpandCase cases[] = {
+      {"substitute string", "foo=bar boo=${foo}", "boo", "bar", hit::Field::Kind::String},
+      {"substute number", "foo=42 boo=${foo}", "boo", "42", hit::Field::Kind::Int},
+      {"multiple replacements",
+       "foo=42 boo=${foo} ${foo}",
+       "boo",
+       "42 42",
+       hit::Field::Kind::String},
+      {"nested", "foo=bar [hello] boo='${foo}' []", "hello/boo", "bar", hit::Field::Kind::String},
+      {"nested shadow",
+       "foo=bar [hello] foo=baz boo='${foo}' []",
+       "hello/boo",
+       "baz",
+       hit::Field::Kind::String},
+  };
+
+  for (size_t i = 0; i < sizeof(cases) / sizeof(ExpandCase); i++)
+  {
+    auto test = cases[i];
+    auto root = hit::parse("TEST", test.input);
+    ExpandWalker exw("TEST");
+    root->walk(&exw);
+    auto n = root->find(test.key);
+    if (!n)
+    {
+      FAIL() << "case " << i + 1 << " failed to find key '" << test.key << "'\n";
+      continue;
+    }
+    if (n->strVal() != test.val)
+    {
+      FAIL() << "case " << i + 1 << " wrong value (key=" << test.key << "): got '" << n->strVal()
+             << "', want '" << test.val << "'\n";
+      continue;
+    }
+
+    auto f = dynamic_cast<hit::Field *>(n);
+    if (f->kind() != test.kind)
```

Comment:
Looks like you routinely use the dynamic_cast, then use the resulting type directly without checking it. Dynamic cast are usually used where the cast may or may not succeed, but in your case you are definitely assuming it does. A static cast might be a better choice.